src/libutil/error.cc

```diff
@@ -50,6 +50,40 @@ std::ostream & operator <<(std::ostream & str, const AbstractPos & pos)
     return str;
 }
 
+inline bool operator<(const AbstractPos& lhs, const AbstractPos& rhs)
+{
+    if (lhs.line != rhs.line)
+        return lhs.line < rhs.line;
+    if (lhs.column != rhs.column)
+        return lhs.column < rhs.column;
+    return false;
+}
+inline bool operator> (const AbstractPos& lhs, const AbstractPos& rhs) { return rhs < lhs; }
+inline bool operator<=(const AbstractPos& lhs, const AbstractPos& rhs) { return !(lhs > rhs); }
+inline bool operator>=(const AbstractPos& lhs, const AbstractPos& rhs) { return !(lhs < rhs); }
+
+inline bool operator<(const Trace& lhs, const Trace& rhs)
+{
+    if (lhs.pos != rhs.pos) {
+        if (!lhs.pos)
+            return true;
+        if (!rhs.pos)
+            return false;
+        return *lhs.pos < *rhs.pos;
+    }
+
+    if (lhs.hint.str() != rhs.hint.str())
+        return lhs.hint.str() < rhs.hint.str();
+
+    if (lhs.frame != rhs.frame)
+        return lhs.frame < rhs.frame;
+
+    return false;
+}
+inline bool operator> (const Trace& lhs, const Trace& rhs) { return rhs < lhs; }
+inline bool operator<=(const Trace& lhs, const Trace& rhs) { return !(lhs > rhs); }
+inline bool operator>=(const Trace& lhs, const Trace& rhs) { return !(lhs < rhs); }
```

Comment:
`std::tie` seems like a great middle ground of having specified semantics, but less code.

Reply:
In the C++20 future, one might want to consider using the three-way comparison operator `<=>`. Its implementation allows the compiler to deduce the rest.
This will have multiple advantages:

- we can get rid of preprocessor macros
- strange semantics that are already implemented individually in the different operator implementation might surface in the light of a unified implementation of the `<=>` op

Reply:
Interesting. Do you know what's blocking this? We do use `c++2a`. Should we track it somewhere?

Reply:
I don't know if anything blocks this at all. Maybe it's just that no one thought about using it, yet.

Reply:
Yes, the `c++2a` addition is recent-ish. At least more recent than the cpp machinery

Reply:
As far as I know Nix is already c++20 and we should probably just write this refactor and make all our comparison code go to space :)

Reply:
>  one of the fields here is shared_ptr which by default has very undesirable comparison semantics (pointer address).

Well, if the only reason we need that comparison operator is putting stuff in a `Set`, pointer comparisons is quite reasonable (and as fast as it can get)

Reply:
Pointer comparisons are not reasonable, because each `Trace` is newly constructed. There's no interning. So it's quite common to have position values that are semantically equal (same line and column number) with different pointer values. I've tried this change and it breaks the tests.

That said, I can simplify these comparisons somewhat and I can replace the `AbstractPos` one with the default.

Reply:
I wonder how costly it would be to intern positions. We quite possibly should do that anyhow, but at a future date.

Reply:
Looks like LLVM only got `<=>` for tuples implemented last year: https://github.com/llvm/llvm-project/issues/50396

Reply:
> I wonder how costly it would be to intern positions. We quite possibly should do that anyhow, but at a future date.

We do have `PosTable`.

This PR's use case is not in the hot path, so we don't need to explore this now I think.